prompb/io/prometheus/client/decoder.go

```diff
@@ -0,0 +1,775 @@
+// Copyright 2025 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package io_prometheus_client //nolint:revive
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"unicode/utf8"
+	"unsafe"
+
+	proto "github.com/gogo/protobuf/proto"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+type MetricStreamingDecoder struct {
+	in    []byte
+	inPos int
+
+	// TODO(bwplotka): Switch to generator/plugin that won't have those fields accessible e.g. OpaqueAPI
+	// We leverage the fact those two don't collide.
+	*MetricFamily // Without Metric, guarded by overridden GetMetric method.
+	*Metric       // Without Label, guarded by overridden GetLabel method.
+
+	mfData      []byte
+	metrics     []pos
+	metricIndex int
+
+	mData  []byte
+	labels []pos
+}
+
+// NewMetricStreamingDecoder returns a Go iterator that unmarshals given protobuf bytes one
+// metric family and metric at the time, allowing efficient streaming.
+//
+// Do not modify MetricStreamingDecoder between iterations as it's reused to save allocations.
+// GetGauge, GetCounter, etc are also cached, which means GetGauge will work for counter
+// if previously gauge was parsed. It's up to the caller to use Type to decide what
+// method to use when checking the value.
+//
+// TODO(bwplotka): io.Reader approach is possible too, but textparse has access to whole scrape for now.
+func NewMetricStreamingDecoder(data []byte) *MetricStreamingDecoder {
+	return &MetricStreamingDecoder{
+		in:           data,
+		MetricFamily: &MetricFamily{},
+		Metric:       &Metric{},
+		metrics:      make([]pos, 0, 100),
+	}
+}
+
+var errInvalidVarint = errors.New("clientpb: invalid varint encountered")
+
+func (m *MetricStreamingDecoder) NextMetricFamily() error {
+	b := m.in[m.inPos:]
+	if len(b) == 0 {
+		return io.EOF
+	}
+	messageLength, varIntLength := proto.DecodeVarint(b) // TODO(bwplotka): Get rid of gogo.
+	if varIntLength == 0 || varIntLength > binary.MaxVarintLen32 {
+		return errInvalidVarint
+	}
+	totalLength := varIntLength + int(messageLength)
+	if totalLength > len(b) {
+		return fmt.Errorf("clientpb: insufficient length of buffer, expected at least %d bytes, got %d bytes", totalLength, len(b))
+	}
+	m.resetMetricFamily()
+	m.mfData = b[varIntLength:totalLength] // TODO: Copy and reuse slice.
```

Comment:
I didn't follow this TODO.  Why do we want to copy?

Reply:
This is due to this sub slice referencing whole metric family buffer. If one would use it e.g. as a cache key or even label strings it would keep the whole memory down.

I will remove this suggestion, I think it's ok here -- the moment we use it as labels to be appended in DB or add a cache key we HAVE to copy and not reuse this anyway. I am concerned with the interface though - it's not very clear what can be reused, what's not, but that's to clean/document later.